<!DOCTYPE html>
<html lang="en"><head>
<link href="../../bioinformatics_6281473.png" rel="icon" type="image/png">
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.4.551">

  <meta name="author" content="Karen Cristine Goncalves, Ph.D.">
  <meta name="dcterms.date" content="2024-02-16">
  <title>Programming classes - Get read counts</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
      }
    pre.numberSource { margin-left: 3em;  padding-left: 4px; }
    div.sourceCode
      { color: #f8f8f2;  }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #f8f8f2; } /* Normal */
    code span.al { color: #f07178; background-color: #2a0f15; font-weight: bold; } /* Alert */
    code span.an { color: #d4d0ab; } /* Annotation */
    code span.at { color: #00e0e0; } /* Attribute */
    code span.bn { color: #d4d0ab; } /* BaseN */
    code span.bu { color: #abe338; } /* BuiltIn */
    code span.cf { color: #ffa07a; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #abe338; } /* Char */
    code span.cn { color: #ffd700; } /* Constant */
    code span.co { color: #f8f8f2; font-style: italic; } /* Comment */
    code span.cv { color: #ffd700; } /* CommentVar */
    code span.do { color: #f8f8f2; } /* Documentation */
    code span.dt { color: #ffa07a; } /* DataType */
    code span.dv { color: #d4d0ab; } /* DecVal */
    code span.er { color: #f07178; text-decoration: underline; } /* Error */
    code span.ex { color: #00e0e0; font-weight: bold; } /* Extension */
    code span.fl { color: #d4d0ab; } /* Float */
    code span.fu { color: #ffa07a; } /* Function */
    code span.im { color: #abe338; } /* Import */
    code span.in { color: #d4d0ab; } /* Information */
    code span.kw { color: #ffa07a; font-weight: bold; } /* Keyword */
    code span.op { color: #ffa07a; } /* Operator */
    code span.ot { color: #00e0e0; } /* Other */
    code span.pp { color: #dcc6e0; } /* Preprocessor */
    code span.re { color: #00e0e0; background-color: #f8f8f2; } /* RegionMarker */
    code span.sc { color: #abe338; } /* SpecialChar */
    code span.ss { color: #abe338; } /* SpecialString */
    code span.st { color: #abe338; } /* String */
    code span.va { color: #00e0e0; } /* Variable */
    code span.vs { color: #abe338; } /* VerbatimString */
    code span.wa { color: #dcc6e0; } /* Warning */
  </style>
  <link rel="stylesheet" href="../../site_libs/revealjs/dist/theme/quarto.css">
  <link href="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-dark">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Get read counts</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Karen Cristine Goncalves, Ph.D. 
</div>
</div>
</div>

  <p class="date">2024-02-16</p>
</section>
<section id="data-processing-before-r" class="slide level2 scrollable">
<h2>Data processing before R</h2>
<p>Short read sequencing transcriptomic data (in the form of fastq files, sometimes compressed as fastq.gz) pass through the some steps before we can proceed to analysis shown in this class.</p>
<p>The pre-processing is typically done in servers (unix/bash language) and includes:</p>
<ol type="1">
<li>Quality assessment and filtering</li>
<li>IN THE ABSENCE OF A REFERENCE: Transcriptome assembly</li>
<li>Read alignment/mapping</li>
<li>Sorting and indexing</li>
</ol>
<p style="color: red">
<strong>The things shown in this class are for alignments onto reference genomes.</strong>
</p>
</section>
<section id="things-to-know" class="slide level2">
<h2>Things to know</h2>
<ul>
<li><p>Keep track how many reads passed the quality filtering and how many were mapped.</p></li>
<li><p>In the case of a new assembly, you can follow steps on <a href="https://github.com/trinityrnaseq/trinityrnaseq/wiki/Trinity-Transcript-Quantification">TrinityRNASeq wiki</a> for transcript quantification.</p></li>
<li><p>Is it possible to do all the processing in R? YES</p>
<ul>
<li><p>Is it worth it? …</p>
<ul>
<li class="fragment">What you will do in R is use it to connect to another program that will do the job. So you end up using more memory (computer gets slower, takes a lot of time).</li>
<li class="fragment">Some programs are not available for windows, so you may install all the packages necessary just to learn that you cannot use them.</li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="data-processing-before-r---detail" class="slide level2 scrollable">
<h2>Data processing before R - detail</h2>
<ol type="1">
<li>Quality assessment and filtering:
<ul>
<li><a href="https://github.com/OpenGene/fastp">fastp</a> (last update: 2023); <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">fastqc</a> (last update: 2023) followed by <a href="https://github.com/usadellab/Trimmomatic">trimmomatic</a> (last update: 2021)</li>
</ul></li>
<li>Creation of a reference transcriptome:
<ul>
<li><a href="https://github.com/trinityrnaseq/trinityrnaseq">trinity</a> (last update: 2023); <a href="https://github.com/aquaskyline/SOAPdenovo-Trans">SOAPdenovo-Trans</a> (last update: 2022); <a href="https://www.bcgsc.ca/resources/software/trans-abyss">Trans-ABySS</a> (last update: 2018); <a href="https://github.com/dzerbino/oases">Oases</a> (last update: 2015); <a href="https://github.com/dzerbino/velvet">Velvet</a> (last update: 2014)</li>
</ul></li>
<li>Read alignment/mapping:
<ul>
<li>Alignment - finds the exact origin of the reads and their differences compared to the reference (like blasting each read against a genome or transcriptome). It is slow - each alignment may take hours depending on the amount of data and size of the reference. These export SAM or BAM files, requiring indexing (step 4):
<ul>
<li><a href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">bowtie2</a> (last update: 2024); <a href="https://github.com/alexdobin/STAR">STAR</a> (last update: 2023); <a href="https://daehwankimlab.github.io/hisat2/">HISAT2</a> (last update: 2020); <a href="https://bio-bwa.sourceforge.net/">bwa</a> (last update: 2017); <a href="https://ccb.jhu.edu/software/tophat/index.shtml">TOPHAT2</a> (last update: 2016)</li>
</ul></li>
<li>Pseudoalignment/Mapping - hyper fast - each alignment takes minutes. They only give the genomic/transcriptomic origin of each read. These export the read count tables directly:
<ul>
<li><a href="https://salmon.readthedocs.io/en/latest/">Salmon</a> (last update: 2023); <a href="https://pachterlab.github.io/kallisto/about">Kallisto</a> (last update: 2019)</li>
</ul></li>
</ul></li>
<li>If the aligner provides the result as SAM or not sorted by genomic coordinates, use <code>samtools view</code> to sort and export into sorted BAM. Then use <code>samtools index</code> to get the <code>bai</code> files.
<ul>
<li>In the end, you should have 2 files per sample/replicate: <code>sampleX_repY_Aligned.sortedByCoord.out.bam</code> and <code>sampleX_repY_Aligned.sortedByCoord.out.bam.bai</code></li>
</ul></li>
</ol>
<div class="callout-info">
<p>Note that if you only want to perform differential expression analysis, GO term enrichment, etc, you do not need to perform read alignment, just mapping suffices.</p>
</div>
</section>
<section id="data-processing-before-r---code-example" class="slide level2">
<h2>Data processing before R - code example</h2>
<p>The code below executes the 2 commands of the program STAR: the creation of the index of the genome and the alignment.</p>
<p>The texts within <code>${}</code> are variable that should be created before hand. This <code>${r1/_R1.fastq}</code> means “remove _R1.fastq” from <code>${r1}</code> (like a <code>gsub</code>)</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a></a><span class="co"># Note that the adapter sequences depend on the type of sequencer</span></span>
<span id="cb1-2"><a></a><span class="ex">fastp</span> <span class="at">-i</span> <span class="va">${raw_R1}</span> <span class="at">-I</span> <span class="va">${raw_R2}</span><span class="dt">\</span></span>
<span id="cb1-3"><a></a> <span class="at">-o</span> <span class="va">${r1}</span> <span class="at">-O</span> <span class="va">${r2}</span><span class="dt">\</span></span>
<span id="cb1-4"><a></a> <span class="at">--qualified_quality_phred</span> 20<span class="dt">\</span></span>
<span id="cb1-5"><a></a> <span class="at">--unqualified_percent_limit</span> 30<span class="dt">\</span></span>
<span id="cb1-6"><a></a> <span class="at">--adapter_sequence</span><span class="op">=</span>TCGTCGGCAGCGTCAGATGTGTATAAGAGACAG<span class="dt">\</span></span>
<span id="cb1-7"><a></a> <span class="at">--adapter_sequence_r2</span><span class="op">=</span>GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG<span class="dt">\</span></span>
<span id="cb1-8"><a></a> <span class="at">--cut_front</span> <span class="at">--cut_front_window_size</span> 3<span class="dt">\</span></span>
<span id="cb1-9"><a></a> <span class="at">--cut_right</span> <span class="at">--cut_right_window_size</span> 4<span class="dt">\</span></span>
<span id="cb1-10"><a></a> <span class="at">--cut_right_mean_quality</span> 15<span class="dt">\</span></span>
<span id="cb1-11"><a></a> <span class="at">--length_required</span> 25<span class="dt">\</span></span>
<span id="cb1-12"><a></a> <span class="at">--html</span> <span class="va">$DIR</span>/fastpReports/<span class="va">${r1</span><span class="op">/</span><span class="ss">_R1.fastq</span><span class="va">}</span>.html<span class="dt">\</span></span>
<span id="cb1-13"><a></a> <span class="at">--thread</span> 8 </span>
<span id="cb1-14"><a></a></span>
<span id="cb1-15"><a></a><span class="co"># Let's decide if we want to map or align</span></span>
<span id="cb1-16"><a></a><span class="va">MODE</span><span class="op">=</span><span class="st">"align"</span></span>
<span id="cb1-17"><a></a></span>
<span id="cb1-18"><a></a><span class="cf">if</span> <span class="kw">[[</span> <span class="va">$MODE</span> <span class="ot">==</span> <span class="st">"align"</span> <span class="kw">]];</span> <span class="cf">then</span></span>
<span id="cb1-19"><a></a>  <span class="ex">STAR</span><span class="dt">\</span></span>
<span id="cb1-20"><a></a>   <span class="at">--runThreadN</span> <span class="va">${NCPUS}</span><span class="dt">\</span></span>
<span id="cb1-21"><a></a>   <span class="at">--runMode</span> genomeGenerate<span class="dt">\</span></span>
<span id="cb1-22"><a></a>   <span class="at">--genomeDir</span> <span class="va">${genomeIdxDIR}</span><span class="dt">\</span></span>
<span id="cb1-23"><a></a>   <span class="at">--genomeFastaFiles</span> <span class="va">${genomeFastaFiles}</span><span class="dt">\</span></span>
<span id="cb1-24"><a></a>   <span class="at">--sjdbGTFfile</span> <span class="va">${genomeGTFFile}</span><span class="dt">\</span></span>
<span id="cb1-25"><a></a>   <span class="at">--sjdbOverhang</span> <span class="va">${readLength}</span></span>
<span id="cb1-26"><a></a>   </span>
<span id="cb1-27"><a></a>  <span class="ex">STAR</span> <span class="at">--genomeDir</span> <span class="va">${indexDIR}</span>/<span class="dt">\</span></span>
<span id="cb1-28"><a></a>   <span class="at">--runThreadN</span> <span class="va">${NCPUS}</span> <span class="dt">\</span></span>
<span id="cb1-29"><a></a>   <span class="at">--readFilesIn</span> <span class="va">${DIR}</span>/clean_reads/<span class="va">${r1}</span> <span class="va">${DIR}</span>/clean_reads/<span class="va">${r2}</span> <span class="dt">\</span></span>
<span id="cb1-30"><a></a>   <span class="at">--outFileNamePrefix</span> <span class="va">${DIR}</span>/alignments/<span class="va">${r1</span><span class="op">/</span><span class="ss">_R1.fastq</span><span class="va">}</span>_<span class="dt">\</span></span>
<span id="cb1-31"><a></a>   <span class="at">--outSAMtype</span> BAM SortedByCoordinate <span class="dt">\</span></span>
<span id="cb1-32"><a></a>   <span class="at">--outSAMunmapped</span> Within</span>
<span id="cb1-33"><a></a></span>
<span id="cb1-34"><a></a><span class="cf">else</span></span>
<span id="cb1-35"><a></a>  <span class="ex">kallisto</span> index <span class="at">-i</span> <span class="va">${genomeIdxDIR}</span> <span class="va">${genomeFastaFiles}</span></span>
<span id="cb1-36"><a></a>  <span class="ex">kallisto</span> quant <span class="at">-i</span> <span class="va">${genomeIdxDIR}</span><span class="dt">\</span></span>
<span id="cb1-37"><a></a>   <span class="at">-o</span> <span class="va">${DIR}</span>/alignments/<span class="va">${r1</span><span class="op">/</span><span class="ss">_R1.fastq</span><span class="va">}</span><span class="dt">\</span></span>
<span id="cb1-38"><a></a>    <span class="va">${DIR}</span>/clean_reads/<span class="va">${r1}</span><span class="dt">\</span></span>
<span id="cb1-39"><a></a>    <span class="va">${DIR}</span>/clean_reads/<span class="va">${r2}</span></span>
<span id="cb1-40"><a></a><span class="cf">fi</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="installing-the-packages" class="slide level2">
<h2>Installing the packages</h2>
<p>If you use the a reference genome for which the annotation is available in Ensembl, you can create your transcript (or 5’/3’-UTR) database while you are waiting for the alignment of your reads.</p>
<p>For this, we need the packages “GenomicFeatures”, “GenomicAlignments”, “biomaRt”, “tidyverse”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a></a>bioconductor_pkgs <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"GenomicFeatures"</span>, <span class="st">"GenomicAlignments"</span>, <span class="st">"biomaRt"</span>, <span class="st">"Rsamtools"</span>)</span>
<span id="cb2-2"><a></a>cran_pkgs <span class="ot">=</span> <span class="st">"tidyverse"</span></span>
<span id="cb2-3"><a></a></span>
<span id="cb2-4"><a></a>devtools<span class="sc">::</span><span class="fu">source_gist</span>(<span class="st">"https://gist.github.com/KarenGoncalves/0db105bceff4ff69547ee25460dda978"</span>)</span>
<span id="cb2-5"><a></a></span>
<span id="cb2-6"><a></a><span class="fu">install_from_dif_sources</span>(</span>
<span id="cb2-7"><a></a>    <span class="at">cran_packages =</span> cran_pkgs,</span>
<span id="cb2-8"><a></a>    <span class="at">bioconductor_packages =</span> bioconductor_pkgs</span>
<span id="cb2-9"><a></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="setting-up-the-scene" class="slide level2">
<h2>Setting up the scene</h2>
<p>There are several biomarts (databases) stored in different Ensembl pages.</p>
<div class="panel-tabset">
<ul id="tabset-1" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-1-1">biomarts</a></li><li><a href="#tabset-1-2">Selecting a biomart dataset</a></li><li><a href="#tabset-1-3">Creating the transcript database</a></li><li><a href="#tabset-1-4">Whole script</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1">
<div class="callout-not">
<p>There is a bacteria Ensembl, but there is not a biomart for it</p>
</div>
<ul>
<li><a href="https://ensembl.org">Animals</a>
<ul>
<li>host: <code>"https://ensembl.org"</code></li>
<li>mart: <code>"ENSEMBL_MART_ENSEMBL"</code> (currently on version 111)</li>
</ul></li>
<li><a href="https://plants.ensembl.org">Plants</a>
<ul>
<li>host: <code>"https://plants.ensembl.org"</code></li>
<li>mart: <code>"plants_mart"</code> (currently on version 58)</li>
</ul></li>
<li><a href="https://fungi.ensembl.org">Fungi</a>
<ul>
<li>host: <code>"https://fungi.ensembl.org"</code></li>
<li>mart: <code>"fungi_mart"</code> (currently on version 58)</li>
</ul></li>
<li><a href="https://protists.ensembl.org">Protists</a>
<ul>
<li>host: <code>"https://protists.ensembl.org"</code></li>
<li>mart: <code>"protists_mart"</code> (currently on version 58)</li>
</ul></li>
</ul>
<p>There are other biomarts in the each Ensembl. You can check their names and versions with</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a></a>host <span class="ot">=</span> <span class="st">"https://plants.ensembl.org"</span></span>
<span id="cb3-2"><a></a><span class="fu">listMarts</span>(<span class="at">host =</span> host)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            biomart                      version
1       plants_mart      Ensembl Plants Genes 58
2 plants_variations Ensembl Plants Variations 58</code></pre>
</div>
</div>
</div>
<div id="tabset-1-2">
<p>Next thing you need is the name of the dataset.</p>
<p>You can check the ones available by running:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a></a>mart <span class="ot">=</span> <span class="fu">useMart</span>(<span class="at">host =</span> host, <span class="at">biomart =</span> <span class="st">"plants_mart"</span>)</span>
<span id="cb5-2"><a></a><span class="fu">listDatasets</span>(mart) <span class="co"># this is a table, so we can filter it</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a></a><span class="co"># If I am working with Arabidopsis thaliana, I can search for it with filter</span></span>
<span id="cb6-2"><a></a><span class="fu">listDatasets</span>(mart) <span class="sc">%&gt;%</span></span>
<span id="cb6-3"><a></a>  <span class="fu">filter</span>(<span class="fu">grepl</span>(<span class="st">"Arabidopsis thaliana"</span>, description))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            dataset                         description version
1 athaliana_eg_gene Arabidopsis thaliana genes (TAIR10)  TAIR10</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a></a><span class="co"># Now I know I have to use the dataset "athaliana_eg_gene"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-1-3">
<p>Now we have all the data we need, we set our variables:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a></a><span class="co"># Edit the following based on what you need</span></span>
<span id="cb9-2"><a></a>biomart <span class="ot">=</span> <span class="st">"plants_mart"</span></span>
<span id="cb9-3"><a></a>dataset <span class="ot">=</span> <span class="st">"athaliana_eg_gene"</span></span>
<span id="cb9-4"><a></a>prefix <span class="ot">=</span> <span class="st">"ensembl_"</span></span>
<span id="cb9-5"><a></a>host <span class="ot">=</span> <span class="st">"https://plants.ensembl.org"</span></span>
<span id="cb9-6"><a></a></span>
<span id="cb9-7"><a></a><span class="co"># The result will be saved as a .RData object, so we do not recalculate this every time.</span></span>
<span id="cb9-8"><a></a>dir <span class="ot">=</span> <span class="st">"./"</span> <span class="co"># current directory</span></span>
<span id="cb9-9"><a></a>outputPath <span class="ot">=</span> <span class="fu">paste0</span>(dir, <span class="st">"Arabidopsis_TxDb.RData"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Then we create a Transcript DataBase (txdb) from Biomart:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a></a><span class="co"># Following takes a while (a few minutes), so it is better to calculate this once than export the result</span></span>
<span id="cb10-2"><a></a>TxDb <span class="ot">&lt;-</span> <span class="fu">makeTxDbFromBiomart</span>(<span class="at">biomart =</span> biomart,</span>
<span id="cb10-3"><a></a>                            <span class="at">dataset =</span> dataset,</span>
<span id="cb10-4"><a></a>                            <span class="at">id_prefix =</span> prefix,</span>
<span id="cb10-5"><a></a>                            <span class="at">host =</span> host)</span>
<span id="cb10-6"><a></a></span>
<span id="cb10-7"><a></a><span class="co"># TxDb is a weird type of data, difficult to access, so we get the information on the transcripts with the function transcriptsBy</span></span>
<span id="cb10-8"><a></a>tx <span class="ot">&lt;-</span> <span class="fu">transcriptsBy</span>(TxDb,<span class="st">"gene"</span>)</span>
<span id="cb10-9"><a></a></span>
<span id="cb10-10"><a></a><span class="co"># Then we save the two databases into the Rdata object.</span></span>
<span id="cb10-11"><a></a><span class="fu">save</span>(TxDb, tx, <span class="at">file =</span> outputPath)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="tabset-1-4">
<p>If you want to use the script we worked on here, just copy it below and change the section “VARIABLES” to fit your needs</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a></a><span class="do">###########################################</span></span>
<span id="cb11-2"><a></a><span class="do">################ VARIABLES ################</span></span>
<span id="cb11-3"><a></a><span class="do">###########################################</span></span>
<span id="cb11-4"><a></a>species <span class="ot">=</span> <span class="st">"Athaliana"</span></span>
<span id="cb11-5"><a></a>biomart <span class="ot">=</span> <span class="st">"plants_mart"</span></span>
<span id="cb11-6"><a></a>dataset <span class="ot">=</span> <span class="st">"athaliana_eg_gene"</span></span>
<span id="cb11-7"><a></a>prefix <span class="ot">=</span> <span class="st">"ensembl_"</span></span>
<span id="cb11-8"><a></a>host <span class="ot">=</span> <span class="st">"https://plants.ensembl.org"</span></span>
<span id="cb11-9"><a></a>dir <span class="ot">=</span> <span class="st">"./output_tables/"</span> </span>
<span id="cb11-10"><a></a>outputPath <span class="ot">=</span> <span class="fu">paste0</span>(dir, species, <span class="st">"_TxDb.RData"</span>)</span>
<span id="cb11-11"><a></a></span>
<span id="cb11-12"><a></a><span class="do">###########################################</span></span>
<span id="cb11-13"><a></a><span class="do">################ Packages #################</span></span>
<span id="cb11-14"><a></a><span class="do">###########################################</span></span>
<span id="cb11-15"><a></a></span>
<span id="cb11-16"><a></a>bioconductor_pkgs <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"GenomicFeatures"</span>, <span class="st">"GenomicAlignments"</span>, <span class="st">"biomaRt"</span>)</span>
<span id="cb11-17"><a></a></span>
<span id="cb11-18"><a></a>devtools<span class="sc">::</span><span class="fu">source_gist</span>(<span class="st">"https://gist.github.com/KarenGoncalves/0db105bceff4ff69547ee25460dda978"</span>)</span>
<span id="cb11-19"><a></a></span>
<span id="cb11-20"><a></a><span class="fu">install_from_dif_sources</span>(</span>
<span id="cb11-21"><a></a>    <span class="at">bioconductor_packages =</span> bioconductor_pkgs</span>
<span id="cb11-22"><a></a>)</span>
<span id="cb11-23"><a></a></span>
<span id="cb11-24"><a></a><span class="do">###########################################</span></span>
<span id="cb11-25"><a></a><span class="do">############# Create databse ##############</span></span>
<span id="cb11-26"><a></a><span class="do">###########################################</span></span>
<span id="cb11-27"><a></a></span>
<span id="cb11-28"><a></a>TxDb <span class="ot">&lt;-</span> <span class="fu">makeTxDbFromBiomart</span>(<span class="at">biomart =</span> biomart,</span>
<span id="cb11-29"><a></a>                            <span class="at">dataset =</span> dataset,</span>
<span id="cb11-30"><a></a>                            <span class="at">id_prefix =</span> prefix,</span>
<span id="cb11-31"><a></a>                            <span class="at">host =</span> host)</span>
<span id="cb11-32"><a></a>tx <span class="ot">&lt;-</span> <span class="fu">transcriptsBy</span>(TxDb,<span class="st">"gene"</span>)</span>
<span id="cb11-33"><a></a></span>
<span id="cb11-34"><a></a><span class="fu">save</span>(TxDb, tx, <span class="at">file =</span> outputPath)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="getting-the-input-files" class="slide level2">
<h2>Getting the input files</h2>
<p>Now that we have the <code>.bam</code>, <code>.bam.bai</code> and the transcript database RData, count the number of reads aligned to each gene.</p>
<p>For the example here, we will use these files (reads from <em>Arabidopsis thaliana</em> Col-0 GFP plants):</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a></a></span>
<span id="cb12-2"><a></a><span class="fu">sapply</span>(<span class="fu">c</span>(<span class="st">"output_tables/counts"</span>, <span class="st">"output_tables/alignments"</span>),</span>
<span id="cb12-3"><a></a>       dir.create)</span>
<span id="cb12-4"><a></a>site <span class="ot">=</span> <span class="st">"https://karengoncalves.github.io/Programming_classes/exampleData/"</span></span>
<span id="cb12-5"><a></a><span class="fu">sapply</span>(<span class="fu">c</span>(<span class="st">"subset_Col_0_GFP_1_Aligned.sortedByCoord.out.bam"</span>,</span>
<span id="cb12-6"><a></a>          <span class="st">"subset_Col_0_GFP_1_Aligned.sortedByCoord.out.bam.bai"</span>),</span>
<span id="cb12-7"><a></a>       \(x) <span class="fu">download.file</span>(</span>
<span id="cb12-8"><a></a>         <span class="at">url =</span> <span class="fu">paste0</span>(site, x), </span>
<span id="cb12-9"><a></a>         <span class="at">destfile =</span> <span class="fu">paste0</span>(<span class="st">"output_tables/alignments"</span>, x))</span>
<span id="cb12-10"><a></a>)    </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="get-the-counts---prepare-the-variables" class="slide level2">
<h2>Get the counts - prepare the variables</h2>
<p>Now that we have the files, we need to inform R where they are stored and where to put the results, in other words, prepare the variables.</p>
<div class="callout callout-note callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<ul>
<li>The basename here is the name of the sample_rep, without the suffix added by the aligner or samtools nor folder names (these are removed with the function <code>basename</code>)</li>
<li>For the example, we will use a single sample, but we will use a loop so the final script can be used fo multiple samples easily</li>
</ul>
</div>
</div>
</div>
<div class="sourceCode" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a></a>bamFiles <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"/output_tables/alignments/"</span>, </span>
<span id="cb13-2"><a></a>                   <span class="fu">list.files</span>(<span class="at">path =</span> <span class="st">"/output_tables/alignments/"</span>)</span>
<span id="cb13-3"><a></a>)</span>
<span id="cb13-4"><a></a>  </span>
<span id="cb13-5"><a></a>baiFiles <span class="ot">&lt;-</span> <span class="fu">paste0</span>(bamFiles, <span class="st">".bai"</span>)</span>
<span id="cb13-6"><a></a>baseNames <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"_Aligned.sortedByCoord.out.bam"</span>, <span class="st">""</span>,</span>
<span id="cb13-7"><a></a>                  <span class="fu">basename</span>(bamFiles))</span>
<span id="cb13-8"><a></a>TxDbPath <span class="ot">&lt;-</span> <span class="st">"./output_tables/Arabidopsis_TxDb.RData"</span></span>
<span id="cb13-9"><a></a><span class="fu">load</span>(TxDbPath)</span>
<span id="cb13-10"><a></a>outPutDir <span class="ot">&lt;-</span> <span class="st">"./output_tables/"</span></span>
<span id="cb13-11"><a></a></span>
<span id="cb13-12"><a></a><span class="co"># Now we load the RData</span></span>
<span id="cb13-13"><a></a><span class="co"># This line tells R where the alignment files for the sample are and how much of them to read at a time</span></span>
<span id="cb13-14"><a></a>bfl <span class="ot">&lt;-</span> <span class="fu">BamFileList</span>(bamFiles, </span>
<span id="cb13-15"><a></a>                   baiFiles, </span>
<span id="cb13-16"><a></a>                   <span class="at">yieldSize=</span><span class="dv">200000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="get-the-counts---overlaps" class="slide level2">
<h2>Get the counts - overlaps</h2>
<p>Summarize overlaps</p>
<div class="panel-tabset">
<ul id="tabset-2" class="panel-tabset-tabby"><li><a data-tabby-default="" href="#tabset-2-1">Code</a></li><li><a href="#tabset-2-2">Explanation</a></li><li><a href="#tabset-2-3">Saving overlaps result</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1">
<p>The code below uses <code>summarizedOverlaps</code> to count the number of alignments (<code>bfl</code> object) falling in each transcript (<code>tx</code> object, transcript database)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a></a>overlaps <span class="ot">&lt;-</span> <span class="fu">summarizeOverlaps</span>(</span>
<span id="cb14-2"><a></a>  <span class="at">features =</span> tx, <span class="co"># genes' coordinates</span></span>
<span id="cb14-3"><a></a>  <span class="at">reads =</span> bfl, <span class="co"># bam and bai files</span></span>
<span id="cb14-4"><a></a>  <span class="at">mode =</span> <span class="st">"Union"</span>, </span>
<span id="cb14-5"><a></a>  <span class="co"># mode can be also "IntersectStrict" or "IntersectNotEmpty"</span></span>
<span id="cb14-6"><a></a>  <span class="at">singleEnd =</span> F, <span class="at">fragments =</span> T, </span>
<span id="cb14-7"><a></a>  <span class="at">ignore.strand =</span> T</span>
<span id="cb14-8"><a></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-2-2">
<p>In this case, we have paired-end reads (singleEnd = F, fragments = T), we ignore the strand information and we use the <a href="https://htseq.readthedocs.io/en/release_0.11.1/count.html">union mode</a>:</p>
<p><img data-src="https://htseq.readthedocs.io/en/release_0.11.1/_images/count_modes.png"></p>
<div class="callout callout-note callout-style-simple">
<div class="callout-body">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-content">
<p>All the bam/bai files are analysed here. So if you have many, R will take very long to run the code. You can increase or decrease the <code>yieldSize</code> in the function <code>BamFileList</code> (previous slide) to use more/less memory and increase/decrease the speed of the code.</p>
</div>
</div>
</div>
</div>
<div id="tabset-2-3">
<p>We save this result, so if there is an issue with the rest of the code, the heavy part of the program gets saved before R stops.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a></a><span class="co"># </span></span>
<span id="cb15-2"><a></a><span class="fu">save</span>(overlaps, <span class="at">file =</span> <span class="fu">paste0</span>(outPutDir, <span class="st">"overlaps.RData"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="get-counts---export-table" class="slide level2">
<h2>Get counts - export table</h2>
<p>Now that we have the overlaps, we can get the counts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a></a><span class="co"># Then we extract the counts slot, transform into a data.frame and export it as a csv file.</span></span>
<span id="cb16-2"><a></a>countAssays <span class="ot">=</span> <span class="fu">assays</span>(overlaps)<span class="sc">$</span>counts  <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb16-3"><a></a><span class="fu">names</span>(countAssays) <span class="ot">&lt;-</span> baseNames</span>
<span id="cb16-4"><a></a></span>
<span id="cb16-5"><a></a><span class="fu">write_csv</span>(<span class="at">x =</span> countAssays, </span>
<span id="cb16-6"><a></a>          <span class="at">file =</span> <span class="fu">paste0</span>(outPutDir, <span class="st">"counts.csv"</span>), </span>
<span id="cb16-7"><a></a>          <span class="at">col_names =</span> T,</span>
<span id="cb16-8"><a></a>          <span class="at">quote =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can check the number of reads/fragments that passed the criteria to be considered mapped using <code>colSums</code>. By using this function, we get the result for all datasets included in the table, without needing to explicitly type any specific sample name.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a></a><span class="fu">colSums</span>(countAssays)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>subset_Col_0_GFP_1 
             51725 </code></pre>
</div>
</div>
</section>
<section id="more-tutorials-and-resources" class="slide level2">
<h2>More tutorials and resources</h2>
<ul>
<li>ARTICLE: <a href="https://doi.org/10.1186/s13059-016-0881-8">a survey of best practices for RNA-seq data analysis</a></li>
<li><a href="https://bioinformatics-core-shared-training.github.io/Bulk_RNAseq_Course_Nov23/Bulk_RNAseq_Course_Base/Markdowns/03_Quantification_with_Salmon_introduction.html">Introduction to transcript quantification with Salmon</a></li>
<li><a href="https://www.alzheimersworkbench.ucsd.edu/EndToEndAnalysis_RNASeq.html">End-to-end analysis - RNAseq</a></li>
<li><a href="https://bioinformatics-core-shared-training.github.io/RNAseq-R/">RNAseq analysis in R - University of Cambridge</a></li>
<li><a href="https://sbc.shef.ac.uk/RNAseq-R/">RNAseq-R</a></li>
</ul>

<div class="quarto-auto-generated-content">
<div class="footer footer-default">

</div>
</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: false,

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    

    <script>

      // htmlwidgets need to know to resize themselves when slides are shown/hidden.

      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current

      // slide changes (different for each slide format).

      (function () {

        // dispatch for htmlwidgets

        function fireSlideEnter() {

          const event = window.document.createEvent("Event");

          event.initEvent("slideenter", true, true);

          window.document.dispatchEvent(event);

        }

    

        function fireSlideChanged(previousSlide, currentSlide) {

          fireSlideEnter();

    

          // dispatch for shiny

          if (window.jQuery) {

            if (previousSlide) {

              window.jQuery(previousSlide).trigger("hidden");

            }

            if (currentSlide) {

              window.jQuery(currentSlide).trigger("shown");

            }

          }

        }

    

        // hookup for slidy

        if (window.w3c_slidy) {

          window.w3c_slidy.add_observer(function (slide_num) {

            // slide_num starts at position 1

            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);

          });

        }

    

      })();

    </script>

    

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp("https:\/\/KarenGoncalves\.github\.io\/Programming_classes\/");
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>